### PR TITLE
Introduce project_name in .hanamirc.

### PR DESCRIPTION
**Pull request by lucasallan**
For more information: https://github.com/hanami/hanami/issues/500

@joneslee85 could you do a quick review/feedback? I might be missing something.

### CODE REVIEW

**Comment by jodosha** [comment]
@lucasallan Yes, you're missing something: the PR mentions `project_name`, but the code has `application_name`. 😉 We should use `project_name`.

**Comment by jodosha** [comment]
@lucasallan I tried to manually test it, but it doesn't work.

```shell
➜ cd Code/hanami/hanami
➜ git status
On branch master
Your branch is up-to-date with 'origin/master'.
nothing to commit, working directory clean
➜ git checkout -b lucasallan-GH-500 master
Switched to a new branch 'lucasallan-GH-500'
➜ git pull https://github.com/lucasallan/lotus.git GH-500
From https://github.com/lucasallan/lotus
 * branch            GH-500     -> FETCH_HEAD
Merge made by the 'recursive' strategy.
 lib/hanami/generators/application/app/hanamirc.tt             |  1 +
 lib/hanami/hanamirc.rb                                        | 59 +++++++++++++++++++++++++++++++++++++++++------------------
 test/fixtures/.hanamirc                                       |  1 +
 test/fixtures/commands/application/new_app/.hanamirc.minitest |  1 +
 test/fixtures/commands/application/new_app/.hanamirc.rspec    |  1 +
 test/fixtures/commands/application/new_app/.hanamirc.slim     |  1 +
 test/fixtures/hanamirc/exists/.hanamirc                       |  3 ++-
 test/fixtures/hanamirc/with_minitest                          |  1 +
 test/fixtures/hanamirc/with_rspec                             |  1 +
 test/hanamirc_test.rb                                         |  3 ++-
 10 files changed, 52 insertions(+), 20 deletions(-)
➜ git status
On branch lucasallan-GH-500
nothing to commit, working directory clean
➜ bundle exec hanami new bookshelf --hanami-head=true
cat bookshelf/.hanamirc
architecture=container
test=minitest
template=erb
```

`project_name` is missing.

File: lib/hanami/hanamirc.rb [hunk 1]
```diff
@@ -158,5 +169,17 @@ def parse_file(path)
...
+    # Generates a default application name based on the application directory
+    #
+    # @since 0.8.0
+    # @api private
+    #
+    # @return [String] application_name
+    #
+    # @see Hanami::Hanamirc::APPLICATION_NAME
+    def default_application_name
+      @app_name || ::File.basename(@root)
```
Comment (jodosha) [inline lib/hanami/hanamirc.rb]: @lucasallan This `@app_name` is always undefined. Where it supposed to be instantiated?

**Comment by jodosha** [comment]
@lucasallan I appended `application_name=bookshelf` to the `.hanamirc` of my test project. Then I renamed the directory of my project from `bookshelf` to `books`. Please note that I used `application_name` to try the current code, but it should be `project_name`.

But if try to generate a new model, it still uses the current directory as project name.

```shell
➜ pwd
/Users/luca/Code/hanami/hanami/books
➜ bundle exec hanami generate model category
      create  lib/books/entities/category.rb
      create  lib/books/repositories/category_repository.rb
      create  spec/books/entities/category_spec.rb
      create  spec/books/repositories/category_repository_spec.rb
```

It uses `books`, but it should've been used `bookshelf` because of the contents in `.hanamirc`.

**Comment by jodosha** [comment]
@lucasallan One last change to make this complete is to hook the `project_name` in model generator.

Specifically, this method `#app_name` should be renamed into `#project_name` and should use the value from `.hanamirc`. https://github.com/hanami/hanami/blob/master/lib/hanami/commands/generate/model.rb#L96

In this way, the directory of the generated models, won't be dependent on the name of the directory, but on the value stored in `.hanamirc`. This will give predictability to the model generator.

**Comment by jodosha** [comment]
@joneslee85 Is there any other place where `project_name` is useful?

**Comment by joneslee85** [comment]
@jodosha no, I could not think of any atm.

**Comment by jodosha** [comment]
@lucasallan @joneslee85 Thanks for taking care of this, I'm gonna to finish this PR myself as I want to wrap a release for the end of this week. Thank you!

**Comment by lucasallan** [comment]
Sorry for the delay @jodosha and thanks for taking care of it.

**Pull request merged**